File: certmonitor/starttls.py
"""Application-protocol preambles that upgrade a plain socket to TLS.

Some services start in plaintext and switch to TLS only after a short
exchange (STARTTLS). Each function here performs that exchange on an
already-connected socket and returns once the server has agreed to start
TLS, leaving the socket ready for `SSLContext.wrap_socket()`. Nothing here
imports beyond the standard library.
"""

from __future__ import annotations

import socket
import struct
from collections.abc import Callable

PROTOCOLS = ("smtp", "imap", "pop3", "ftp", "postgres", "ldap")

LDAP_STARTTLS_OID = b"1.3.6.1.4.1.1466.20037"
_POSTGRES_SSL_REQUEST_CODE = 80877103
_LINE_LIMIT = 4096


class StartTLSError(OSError):
    """The server did not agree to start TLS, or the preamble was malformed."""


def negotiate(
    sock: socket.socket, protocol: str, *, client_name: str = "certmonitor"
) -> None:
    """Run the STARTTLS preamble for `protocol` on `sock`.

    Args:
        sock: A connected plaintext socket.
        protocol: One of `PROTOCOLS`.
        client_name: Name announced to servers that ask for one (SMTP EHLO).

    Raises:
        ValueError: If `protocol` is not supported.
        StartTLSError: If the server refuses or answers unexpectedly.
    """
    handler = _HANDLERS.get(protocol)
    if handler is None:
        raise ValueError(
            f"unsupported STARTTLS protocol {protocol!r}; choose one of {', '.join(PROTOCOLS)}"
        )
    handler(sock, client_name)


# --- line-oriented protocols -------------------------------------------------


def _read_line(sock: socket.socket) -> str:
    """Read one CRLF- or LF-terminated line, byte by byte so nothing past it is consumed."""
    line = bytearray()
    while not line.endswith(b"\n"):
        byte = sock.recv(1)
        if not byte:
            raise StartTLSError("connection closed during STARTTLS negotiation")
        line += byte
        if len(line) > _LINE_LIMIT:
            raise StartTLSError("STARTTLS reply line too long")
    return line.decode("utf-8", errors="replace").rstrip("\r\n")


def _read_reply(sock: socket.socket) -> tuple[str, list[str]]:
    """Read an SMTP/FTP style reply; multiline replies use `NNN-` continuation lines."""
    lines: list[str] = []
    while True:
        line = _read_line(sock)
        lines.append(line)
        if len(line) < 4 or line[3] != "-":
            return line[:3], lines


def _expect(sock: socket.socket, code: str, what: str) -> list[str]:
    got, lines = _read_reply(sock)
    if got != code:
        raise StartTLSError(f"{what}: expected {code}, server said {lines[-1]!r}")
    return lines


def _smtp(sock: socket.socket, client_name: str) -> None:
    _expect(sock, "220", "SMTP greeting")
    sock.sendall(f"EHLO {client_name}\r\n".encode())
    capabilities = _expect(sock, "250", "SMTP EHLO")
    if not any(line[4:].upper().startswith("STARTTLS") for line in capabilities):
        raise StartTLSError("SMTP server does not advertise STARTTLS")
    sock.sendall(b"STARTTLS\r\n")
    _expect(sock, "220", "SMTP STARTTLS")


def _ftp(sock: socket.socket, client_name: str) -> None:
    _expect(sock, "220", "FTP greeting")
    sock.sendall(b"AUTH TLS\r\n")
    _expect(sock, "234", "FTP AUTH TLS")


def _imap(sock: socket.socket, client_name: str) -> None:
    greeting = _read_line(sock)
    if not greeting.upper().startswith("* OK") and not greeting.upper().startswith(
        "* PREAUTH"
    ):
        raise StartTLSError(f"IMAP greeting: server said {greeting!r}")
    sock.sendall(b"a001 STARTTLS\r\n")
    while True:
        line = _read_line(sock)
        if line.startswith("a001 "):
            if line[5:].upper().startswith("OK"):
                return
            raise StartTLSError(f"IMAP STARTTLS: server said {line!r}")


def _pop3(sock: socket.socket, client_name: str) -> None:
    greeting = _read_line(sock)
    if not greeting.startswith("+OK"):
        raise StartTLSError(f"POP3 greeting: server said {greeting!r}")
    sock.sendall(b"STLS\r\n")
    reply = _read_line(sock)
    if not reply.startswith("+OK"):
        raise StartTLSError(f"POP3 STLS: server said {reply!r}")


# --- binary protocols --------------------------------------------------------


def _recv_exact(sock: socket.socket, size: int) -> bytes:
    data = bytearray()
    while len(data) < size:
        chunk = sock.recv(size - len(data))
        if not chunk:
            raise StartTLSError("connection closed during STARTTLS negotiation")
        data += chunk
    return bytes(data)


def _postgres(sock: socket.socket, client_name: str) -> None:
    """PostgreSQL SSLRequest: 8 bytes out, one byte back ('S' agrees, 'N' declines)."""
    sock.sendall(struct.pack("!ii", 8, _POSTGRES_SSL_REQUEST_CODE))
    answer = _recv_exact(sock, 1)
    if answer == b"S":
        return
    if answer == b"N":
        raise StartTLSError("PostgreSQL server declined SSL")
    raise StartTLSError(f"PostgreSQL SSLRequest: unexpected reply {answer!r}")


def _ber(tag: int, content: bytes) -> bytes:
    """Encode one BER TLV with a definite length."""
    length = len(content)
    if length < 0x80:
        header = bytes([tag, length])
    else:
        size = (length.bit_length() + 7) // 8
        header = bytes([tag, 0x80 | size]) + length.to_bytes(size, "big")
    return header + content


def _ber_read(data: bytes, offset: int) -> tuple[int, bytes, int]:
    """Decode one BER TLV at `offset`; return (tag, content, next offset)."""
    if offset + 2 > len(data):
        raise StartTLSError("LDAP reply truncated")
    tag, first = data[offset], data[offset + 1]
    offset += 2
    if first & 0x80:
        size = first & 0x7F
        if size == 0 or offset + size > len(data):
            raise StartTLSError("LDAP reply has an unsupported length encoding")
        length = int.from_bytes(data[offset : offset + size], "big")
        offset += size
    else:
        length = first
    if offset + length > len(data):
        raise StartTLSError("LDAP reply truncated")
    return tag, data[offset : offset + length], offset + length


def ldap_starttls_request(message_id: int = 1) -> bytes:
    """The LDAPMessage carrying an ExtendedRequest for the StartTLS OID (RFC 4511)."""
    extended_request = _ber(
        0x77, _ber(0x80, LDAP_STARTTLS_OID)
    )  # [APPLICATION 23], [0] requestName
    return _ber(0x30, _ber(0x02, bytes([message_id])) + extended_request)


def _ldap(sock: socket.socket, client_name: str) -> None:
    sock.sendall(ldap_starttls_request())
    # Read the outer SEQUENCE header first so exactly one message is consumed.
    head = _recv_exact(sock, 2)
    if head[0] != 0x30:
        raise StartTLSError(
            f"LDAP reply: expected an LDAPMessage, got tag 0x{head[0]:02x}"
        )
    if head[1] & 0x80:
        size = head[1] & 0x7F
        if size == 0 or size > 4:
            raise StartTLSError("LDAP reply has an unsupported length encoding")
        length = int.from_bytes(_recv_exact(sock, size), "big")
    else:
        length = head[1]
    body = _recv_exact(sock, length)
    _, _message_id, offset = _ber_read(body, 0)
    op_tag, op, _ = _ber_read(body, offset)
    if op_tag != 0x78:  # [APPLICATION 24] ExtendedResponse
        raise StartTLSError(
            f"LDAP reply: expected an ExtendedResponse, got tag 0x{op_tag:02x}"
        )
    _, result_code, offset = _ber_read(op, 0)
    _, _matched_dn, offset = _ber_read(op, offset)
    _, diagnostic, _ = _ber_read(op, offset)
    code = int.from_bytes(result_code, "big") if result_code else -1
    if code != 0:
        detail = diagnostic.decode("utf-8", errors="replace") or f"resultCode {code}"
        raise StartTLSError(f"LDAP StartTLS refused: {detail}")


_HANDLERS: dict[str, Callable[[socket.socket, str], None]] = {
    "smtp": _smtp,
    "imap": _imap,
    "pop3": _pop3,
    "ftp": _ftp,
    "postgres": _postgres,
    "ldap": _ldap,
}

File: rust_certinfo/src/tls/starttls.rs
// rust_certinfo/src/tls/starttls.rs
//
// STARTTLS preambles for the PQ probe, mirroring certmonitor/starttls.py:
// SMTP, IMAP, POP3, FTP, PostgreSQL, and LDAP. Each function drives an
// already-connected TcpStream through the plaintext exchange and returns
// once the server has agreed to start TLS, leaving the stream ready for
// the ClientHello. Line-oriented protocols are read one byte at a time so
// nothing past the server's final reply is consumed. Every read and write
// is bounded by the probe's deadline.

use std::net::TcpStream;
use std::time::Instant;

use crate::tls::wire;

pub const PROTOCOLS: [&str; 6] = ["smtp", "imap", "pop3", "ftp", "postgres", "ldap"];
const LDAP_STARTTLS_OID: &[u8] = b"1.3.6.1.4.1.1466.20037";
const POSTGRES_SSL_REQUEST: [u8; 8] = [0, 0, 0, 8, 0x04, 0xd2, 0x16, 0x2f];

/// Run the STARTTLS preamble for `protocol`. The error string is the
/// server's reply where there is one.
pub fn negotiate(stream: &mut TcpStream, protocol: &str, deadline: Instant) -> Result<(), String> {
    match protocol {
        "smtp" => smtp(stream, deadline),
        "imap" => imap(stream, deadline),
        "pop3" => pop3(stream, deadline),
        "ftp" => ftp(stream, deadline),
        "postgres" => postgres(stream, deadline),
        "ldap" => ldap(stream, deadline),
        other => Err(format!(
            "unsupported STARTTLS protocol {other:?}; choose one of {}",
            PROTOCOLS.join(", ")
        )),
    }
}

// --- I/O helpers ---------------------------------------------------------------------

const WHAT: &str = "STARTTLS negotiation";

fn send(stream: &mut TcpStream, bytes: &[u8], deadline: Instant) -> Result<(), String> {
    wire::send(stream, bytes, deadline, WHAT)
}

fn read_exact(stream: &mut TcpStream, size: usize, deadline: Instant) -> Result<Vec<u8>, String> {
    wire::read_exact(stream, size, deadline, WHAT)
}

fn read_line(stream: &mut TcpStream, deadline: Instant) -> Result<String, String> {
    wire::read_line(stream, deadline, WHAT)
}

/// SMTP/FTP style reply: `NNN-` lines continue, `NNN ` (or short) ends it.
fn read_reply(stream: &mut TcpStream, deadline: Instant) -> Result<(String, Vec<String>), String> {
    let mut lines = Vec::new();
    loop {
        let line = read_line(stream, deadline)?;
        let done = line.len() < 4 || line.as_bytes()[3] != b'-';
        lines.push(line);
        if done {
            let code = lines
                .last()
                .map(|l| l.chars().take(3).collect())
                .unwrap_or_default();
            return Ok((code, lines));
        }
    }
}

fn expect(
    stream: &mut TcpStream,
    code: &str,
    what: &str,
    deadline: Instant,
) -> Result<Vec<String>, String> {
    let (got, lines) = read_reply(stream, deadline)?;
    if got != code {
        let last = lines.last().cloned().unwrap_or_default();
        return Err(format!("{what}: expected {code}, server said {last:?}"));
    }
    Ok(lines)
}

fn after_prefix(line: &str, prefix_len: usize) -> String {
    line.get(prefix_len..).unwrap_or("").to_ascii_uppercase()
}

// --- protocols -------------------------------------------------------------------------

fn smtp(stream: &mut TcpStream, deadline: Instant) -> Result<(), String> {
    expect(stream, "220", "SMTP greeting", deadline)?;
    send(stream, b"EHLO certmonitor\r\n", deadline)?;
    let capabilities = expect(stream, "250", "SMTP EHLO", deadline)?;
    if !capabilities
        .iter()
        .any(|l| after_prefix(l, 4).starts_with("STARTTLS"))
    {
        return Err("SMTP server does not advertise STARTTLS".into());
    }
    send(stream, b"STARTTLS\r\n", deadline)?;
    expect(stream, "220", "SMTP STARTTLS", deadline)?;
    Ok(())
}

fn ftp(stream: &mut TcpStream, deadline: Instant) -> Result<(), String> {
    expect(stream, "220", "FTP greeting", deadline)?;
    send(stream, b"AUTH TLS\r\n", deadline)?;
    expect(stream, "234", "FTP AUTH TLS", deadline)?;
    Ok(())
}

fn imap(stream: &mut TcpStream, deadline: Instant) -> Result<(), String> {
    let greeting = read_line(stream, deadline)?;
    let upper = greeting.to_ascii_uppercase();
    if !upper.starts_with("* OK") && !upper.starts_with("* PREAUTH") {
        return Err(format!("IMAP greeting: server said {greeting:?}"));
    }
    send(stream, b"a001 STARTTLS\r\n", deadline)?;
    loop {
        let line = read_line(stream, deadline)?;
        if line.starts_with("a001 ") {
            if after_prefix(&line, 5).starts_with("OK") {
                return Ok(());
            }
            return Err(format!("IMAP STARTTLS: server said {line:?}"));
        }
    }
}

fn pop3(stream: &mut TcpStream, deadline: Instant) -> Result<(), String> {
    let greeting = read_line(stream, deadline)?;
    if !greeting.starts_with("+OK") {
        return Err(format!("POP3 greeting: server said {greeting:?}"));
    }
    send(stream, b"STLS\r\n", deadline)?;
    let reply = read_line(stream, deadline)?;
    if !reply.starts_with("+OK") {
        return Err(format!("POP3 STLS: server said {reply:?}"));
    }
    Ok(())
}

fn postgres(stream: &mut TcpStream, deadline: Instant) -> Result<(), String> {
    send(stream, &POSTGRES_SSL_REQUEST, deadline)?;
    let answer = read_exact(stream, 1, deadline)?;
    match answer[0] {
        b'S' => Ok(()),
        b'N' => Err("PostgreSQL server declined SSL".into()),
        other => Err(format!(
            "PostgreSQL SSLRequest: unexpected reply {:?}",
            other as char
        )),
    }
}

// --- LDAP (RFC 4511 StartTLS extended operation) --------------------------------------

fn ber(tag: u8, content: &[u8]) -> Vec<u8> {
    let mut out = vec![tag];
    let length = content.len();
    if length < 0x80 {
        out.push(length as u8);
    } else {
        let bytes = length.to_be_bytes();
        let start = bytes
            .iter()
            .position(|b| *b != 0)
            .unwrap_or(bytes.len() - 1);
        out.push(0x80 | (bytes.len() - start) as u8);
        out.extend_from_slice(&bytes[start..]);
    }
    out.extend_from_slice(content);
    out
}

fn ber_read(data: &[u8], offset: usize) -> Result<(u8, &[u8], usize), String> {
    if offset + 2 > data.len() {
        return Err("LDAP reply truncated".into());
    }
    let tag = data[offset];
    let first = data[offset + 1];
    let mut cursor = offset + 2;
    let length = if first & 0x80 != 0 {
        let size = (first & 0x7f) as usize;
        if size == 0 || size > 4 || cursor + size > data.len() {
            return Err("LDAP reply has an unsupported length encoding".into());
        }
        let mut value = 0usize;
        for byte in &data[cursor..cursor + size] {
            value = (value << 8) | *byte as usize;
        }
        cursor += size;
        value
    } else {
        first as usize
    };
    if cursor + length > data.len() {
        return Err("LDAP reply truncated".into());
    }
    Ok((tag, &data[cursor..cursor + length], cursor + length))
}

pub fn ldap_starttls_request() -> Vec<u8> {
    let request_name = ber(0x80, LDAP_STARTTLS_OID);
    let extended_request = ber(0x77, &request_name);
    let mut body = ber(0x02, &[1]);
    body.extend_from_slice(&extended_request);
    ber(0x30, &body)
}

fn ldap(stream: &mut TcpStream, deadline: Instant) -> Result<(), String> {
    send(stream, &ldap_starttls_request(), deadline)?;
    let head = read_exact(stream, 2, deadline)?;
    if head[0] != 0x30 {
        return Err(format!(
            "LDAP reply: expected an LDAPMessage, got tag 0x{:02x}",
            head[0]
        ));
    }
    let length = if head[1] & 0x80 != 0 {
        let size = (head[1] & 0x7f) as usize;
        if size == 0 || size > 4 {
            return Err("LDAP reply has an unsupported length encoding".into());
        }
        read_exact(stream, size, deadline)?
            .iter()
            .fold(0usize, |acc, b| (acc << 8) | *b as usize)
    } else {
        head[1] as usize
    };
    let body = read_exact(stream, length, deadline)?;
    let (_, _message_id, offset) = ber_read(&body, 0)?;
    let (op_tag, op, _) = ber_read(&body, offset)?;
    if op_tag != 0x78 {
        return Err(format!(
            "LDAP reply: expected an ExtendedResponse, got tag 0x{op_tag:02x}"
        ));
    }
    let (_, result_code, offset) = ber_read(op, 0)?;
    let (_, _matched_dn, offset) = ber_read(op, offset)?;
    let (_, diagnostic, _) = ber_read(op, offset)?;
    let code = result_code
        .iter()
        .fold(0usize, |acc, b| (acc << 8) | *b as usize);
    if code != 0 || result_code.is_empty() {
        let detail = String::from_utf8_lossy(diagnostic);
        let detail = if detail.is_empty() {
            format!("resultCode {code}")
        } else {
            detail.to_string()
        };
        return Err(format!("LDAP StartTLS refused: {detail}"));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::{BufRead, BufReader, Read, Write};
    use std::net::TcpListener;
    use std::time::Duration;

    fn serve<F>(handler: F) -> u16
    where
        F: FnOnce(TcpStream) + Send + 'static,
    {
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        let port = listener.local_addr().unwrap().port();
        std::thread::spawn(move || {
            if let Ok((stream, _)) = listener.accept() {
                handler(stream);
            }
        });
        port
    }

    fn connect(port: u16) -> TcpStream {
        TcpStream::connect(("127.0.0.1", port)).unwrap()
    }

    fn deadline() -> Instant {
        Instant::now() + Duration::from_secs(2)
    }

    fn line(reader: &mut BufReader<TcpStream>) -> String {
        let mut s = String::new();
        reader.read_line(&mut s).unwrap();
        s
    }

    #[test]
    fn smtp_preamble_succeeds() {
        let port = serve(|mut s| {
            let mut r = BufReader::new(s.try_clone().unwrap());
            s.write_all(b"220 fake\r\n").unwrap();
            assert!(line(&mut r).starts_with("EHLO "));
            s.write_all(b"250-fake\r\n250-STARTTLS\r\n250 OK\r\n")
                .unwrap();
            assert_eq!(line(&mut r).trim(), "STARTTLS");
            s.write_all(b"220 go\r\n").unwrap();
        });
        assert_eq!(negotiate(&mut connect(port), "smtp", deadline()), Ok(()));
    }

    #[test]
    fn smtp_refusal_carries_the_reply() {
        let port = serve(|mut s| {
            let mut r = BufReader::new(s.try_clone().unwrap());
            s.write_all(b"220 fake\r\n").unwrap();
            line(&mut r);
            s.write_all(b"250 fake\r\n").unwrap();
        });
        let err = negotiate(&mut connect(port), "smtp", deadline()).unwrap_err();
        assert!(err.contains("does not advertise STARTTLS"), "{err}");
    }

    #[test]
    fn imap_pop3_ftp_preambles_succeed() {
        let imap_port = serve(|mut s| {
            let mut r = BufReader::new(s.try_clone().unwrap());
            s.write_all(b"* OK ready\r\n").unwrap();
            assert_eq!(line(&mut r).trim(), "a001 STARTTLS");
            s.write_all(b"* noise\r\na001 OK begin\r\n").unwrap();
        });
        assert_eq!(
            negotiate(&mut connect(imap_port), "imap", deadline()),
            Ok(())
        );
        let pop3_port = serve(|mut s| {
            let mut r = BufReader::new(s.try_clone().unwrap());
            s.write_all(b"+OK\r\n").unwrap();
            assert_eq!(line(&mut r).trim(), "STLS");
            s.write_all(b"+OK begin\r\n").unwrap();
        });
        assert_eq!(
            negotiate(&mut connect(pop3_port), "pop3", deadline()),
            Ok(())
        );
        let ftp_port = serve(|mut s| {
            let mut r = BufReader::new(s.try_clone().unwrap());
            s.write_all(b"220-hi\r\n220 fake\r\n").unwrap();
            assert_eq!(line(&mut r).trim(), "AUTH TLS");
            s.write_all(b"234 ok\r\n").unwrap();
        });
        assert_eq!(negotiate(&mut connect(ftp_port), "ftp", deadline()), Ok(()));
    }

    #[test]
    fn postgres_accepts_and_declines() {
        let ok_port = serve(|mut s| {
            let mut buf = [0u8; 8];
            s.read_exact(&mut buf).unwrap();
            assert_eq!(buf, POSTGRES_SSL_REQUEST);
            s.write_all(b"S").unwrap();
        });
        assert_eq!(
            negotiate(&mut connect(ok_port), "postgres", deadline()),
            Ok(())
        );
        let no_port = serve(|mut s| {
            let mut buf = [0u8; 8];
            s.read_exact(&mut buf).unwrap();
            s.write_all(b"N").unwrap();
        });
        let err = negotiate(&mut connect(no_port), "postgres", deadline()).unwrap_err();
        assert!(err.contains("declined SSL"), "{err}");
    }

    #[test]
    fn ldap_request_bytes_and_replies() {
        let request = ldap_starttls_request();
        assert_eq!(&request[..2], &[0x30, 0x1d]);
        assert_eq!(&request[5..9], &[0x77, 0x18, 0x80, 0x16]);
        let ok_port = serve(|mut s| {
            let mut buf = vec![0u8; 31];
            s.read_exact(&mut buf).unwrap();
            assert_eq!(buf, ldap_starttls_request());
            s.write_all(&[
                0x30, 0x0c, 0x02, 0x01, 0x01, 0x78, 0x07, 0x0a, 0x01, 0x00, 0x04, 0x00, 0x04, 0x00,
            ])
            .unwrap();
        });
        assert_eq!(negotiate(&mut connect(ok_port), "ldap", deadline()), Ok(()));
        let refuse_port = serve(|mut s| {
            let mut buf = vec![0u8; 31];
            s.read_exact(&mut buf).unwrap();
            let mut reply = vec![
                0x30, 0x1d, 0x02, 0x01, 0x01, 0x78, 0x18, 0x0a, 0x01, 0x01, 0x04, 0x00, 0x04, 0x11,
            ];
            reply.extend_from_slice(b"TLS not supported");
            s.write_all(&reply).unwrap();
        });
        let err = negotiate(&mut connect(refuse_port), "ldap", deadline()).unwrap_err();
        assert!(err.contains("TLS not supported"), "{err}");
    }

    #[test]
    fn closed_connection_and_unknown_protocol_are_errors() {
        let port = serve(|mut s| {
            s.write_all(b"220 fake\r\n").unwrap();
        });
        let err = negotiate(&mut connect(port), "smtp", deadline()).unwrap_err();
        assert!(
            err.contains("connection closed") || err.contains("read failed"),
            "{err}"
        );
        let port = serve(|_s| {});
        let err = negotiate(&mut connect(port), "gopher", deadline()).unwrap_err();
        assert!(err.contains("unsupported STARTTLS protocol"), "{err}");
    }

    #[test]
    fn ber_long_form_round_trips() {
        let payload = vec![b'x'; 300];
        let encoded = ber(0x04, &payload);
        assert_eq!(&encoded[..4], &[0x04, 0x82, 0x01, 0x2c]);
        let (tag, content, next) = ber_read(&encoded, 0).unwrap();
        assert_eq!((tag, content.len(), next), (0x04, 300, encoded.len()));
        assert!(ber_read(&[0x04], 0).is_err());
        assert!(ber_read(&[0x04, 0x80], 0).is_err());
        assert!(ber_read(&[0x04, 0x05, b'a'], 0).is_err());
    }
}
